3단계 - 로또(자동) 코드 리뷰 부탁드립니다.

grow-with opened the pull request:
안녕하세요
드디어 3단계를 보냅니다 ㅠ
목요일에 강의 듣고 원시값 포장와 일급콜렉션을 만드는 것에 대해서 엄청나게 고민을 많이 한것 같습니다.
테스트 코드부터 코드를 짜서 나아갈려고 생각을 엄청 많이 했던 것 같습니다.
리뷰 해주셨던 내용을 놓치지 않으려고 클래스의 책임에 대해서 많이 고민하고 작성하였습니다.
코드를 작성하면서 들었던 의문이 몇가지 있어서 질문 드려봅니다. 

1. 1급 컬렉션을 만들 때 내부에 상수를 정의 해도 되나요?
   (추후에 공통 상수와 메세지는 전부 외부로 취합해주었는데 해당 1급 컬렉션만이 가지는 특별한 의미의 상수들은 내부에 같이 선언을 해도 괜찮을까요?)
2. 테스트코드 작성을 하기 위해서 준비작업이 좀 발생해도 괜찮나요? (ex String을 split해서 List<Integer)로 만든다거나)(
   (괜히 테스트 코드의 given이 너무 많아 지면 가독성이 안좋을 까봐 걱정되서 질문해봅니다.)
3. 메서드 내부의 지역변수로 사용하기 위해 선언하는 원시값들을 보통 0이나 false등과 같은 걸로 초기화하는데 해당 부분들도 상수로 정의를 해주는게 좋은가요? 뭔가 상수로 정의하는것이 더 가독성이 안좋은 느낌이 들어서 여쭤봅니다.
4. 원시포장 객체나 1급 컬렉션의 경우 final을 선언해서 재할당을 금지하게 만드는게 좋다고 알고는 있는데 막상 작성하다보면 해당 내부의 값에 대해서 변경이 일어나야 하는 경우가 존재하는데 이런 경우에도 final을 최대한 유지하는 방향으로 코드작성을 생각해봐야할까요?? 아니면 상황에 타협해서 사용해도 될까요?

프리코스를 거치면서 그래도 해봤던 부분이라 수월할 줄 알았는데 TDD라는게 정말 어렵다는걸 이번 미션에서 느낍니다 ㅠ
거의 2일을 고민해가면서 작성하였지만 부족한 부분이 많을거라 생각됩니다.
리뷰해주시느라 항상 수고가 많으십니다.
좋은 리뷰 항상 감사드립니다.

grow-with committed:
docs: 3단계 기능 요구사항 정리

grow-with committed:
docs: 3단계 기능 요구사항 수정

grow-with committed:
feat: 로또번호 6자리 생성 기능 구현

- 테스트 코드 작성
  6개의 번호가 생성되었는지 확인
  생성된 6개의 번호가 중복이 있는지 확인

grow-with committed:
feat: 당첨번호 일치하는 개수 확인 기능 구현

- 테스트 코드 작성
  번호 1개 일치 여부 확인 기능 테스트
  번호 6개 3개 0개 기능 테스트

grow-with committed:
feat: 당첨통계 계산 기능 구현

- 테스트 코드 작성
  등수 별 당첨자수 확인
  수익률 확인

grow-with committed:
feat: 구입금액을 입력받는 기능 구현

- 숫자 외의 입력값을 받으면 에러 발생

grow-with committed:
feat: 구매 수량을 출력하는 기능구현

grow-with committed:
feat: 구매한 로또들의 번호를 출력하는 기능 구현

grow-with committed:
feat: 당첨번호 입력받기 기능 구현

grow-with committed:
docs: 기능 요구사항 수정

grow-with committed:
feat: 당첨 통계 출력 기능 구현

grow-with committed:
refactor: 공통 상수 및 메세지 정리

grow-with committed:
docs: 기능 요구사항 수정

leeyohan93 commented on README.md:
```diff
@@ -1,9 +1,25 @@
 # 로또
```
>1급 컬렉션을 만들 때 내부에 상수를 정의 해도 되나요?
(추후에 공통 상수와 메세지는 전부 외부로 취합해주었는데 해당 1급 컬렉션만이 가지는 특별한 의미의 상수들은 내부에 같이 선언을 해도 괜찮을까요?)

상수는 보통 static final로 선언 되는데, 상수의 초기화 시점은 JVM에 의해 클래스가 메모리에 올라가는 순간이며,
저희가 실질적으로 일급컬렉션의 인스턴스 즉, 객체에는 아무런 영향을 주지 않기 때문에 상수는 얼마든지 정의해도 좋다고 생각합니다 :) 상수를 내부에 정의하게되면 무분별한 상수 클래스가 만들어지지 않게되고 상수의 의미가 더욱 명확해 질 수 있지 않을까요 ? (ex. Lotto.PRICE)

>테스트코드 작성을 하기 위해서 준비작업이 좀 발생해도 괜찮나요? (ex String을 split해서 List<Integer)로 만든다거나)(
(괜히 테스트 코드의 given이 너무 많아 지면 가독성이 안좋을 까봐 걱정되서 질문해봅니다.)

테스트의 가독성을 챙겨주시는 것이 너무 좋네요 👍 👍  
테스트 코드 작성 시 준비 작업이 발생할 수 밖에 없다고 생각합니다.
준비 작업을 위한 @BeforeEach, @MethodSource, Test Fixture 사용 등 여러 가지 준비 작업을 위한 어노테이션이나 전략이 있는데요. 상황에 맞게 사용하여 테스트의 가독성을 올리는 것이 중요합니다.

> 메서드 내부의 지역변수로 사용하기 위해 선언하는 원시값들을 보통 0이나 false등과 같은 걸로 초기화하는데 해당 부분들도 상수로 정의를 해주는게 좋은가요? 뭔가 상수로 정의하는것이 더 가독성이 안좋은 느낌이 들어서 여쭤봅니다.

저는 개인적으로 보통 이런 경우가 있을때는, 상수로 정의해주지 않는 것이 낫다고 생각하는데요.
상수는 의미 있는 어떤 값을 표현할 때 더욱 적합하다고 생각하기 때문입니다 :)

> 원시포장 객체나 1급 컬렉션의 경우 final을 선언해서 재할당을 금지하게 만드는게 좋다고 알고는 있는데 막상 작성하다보면 해당 내부의 값에 대해서 변경이 일어나야 하는 경우가 존재하는데 이런 경우에도 final을 최대한 유지하는 방향으로 코드작성을 생각해봐야할까요?? 아니면 상황에 타협해서 사용해도 될까요?

개인적으로는 어쩔 수 없는 상황(ex. JPA의 엔티티..)을 제외하고는 final을 유지하는 방향을 권해드리고 싶어요. 
그 이유는 불변 객체가 가지는 여러 장점들로 설명드릴 수 있겠지만 그 중 신뢰성과 관려된 이야기를 드리자면,
재할당이나 내부의 값에 대해 변경이 일어난다면 우리는 해당 객체를 사용하기 위해 다시 검증하고 언제 일어날지 모를 사고를 예방해야합니다. 모든 코드에 이러한 사고 예방은 쉽지 않은 일이라고 생각해요 :)

leeyohan93 commented on README.md:
```diff
@@ -1,9 +1,25 @@
+
+---
+
+### 3단계 - 로또(자동) 기능 요구사항 정리
```
기능 요구사항 잘 작성해주셨습니다 ! 👍

이에 더해 요구사항에 도메인 모델들을 도출하는 단계까지 하실 수 있을 것 같아요 :)

코드 구현 이전에 도메인 모델을 도출하고 구현해야할 기능들을 책임에 맞게 할당해보는 것도 또 다른 좋은 설계 방법인 것 같아요 !

(포비의 말씀처럼 도메인에 대한 이해가 없다면 일단 코드 구현! 그리고 코드 구현을 모두 날리고 다시 시작하는 방법도 좋습니다)

leeyohan93 commented on src/main/java/lotto/CommonConstant.java:
```diff
@@ -0,0 +1,23 @@
+package lotto;
+
+public class CommonConstant {
```
상수에 대해 코멘트 드린 부분을 고민해주세요 😄

leeyohan93 commented on src/main/java/lotto/CommonMessage.java:
```diff
@@ -0,0 +1,13 @@
+	public static final String PRINT_PROFIT_RATE = "총 수익률은 %.2f입니다.";
+	public static final String ERROR_PURCHASE_AMOUNT_TYPE = "[ERROR] 구매금액은 숫자만 입력가능합니다.";
+	public static final String ERROR_PURCHASE_AMOUNT = "[ERROR] 구매금액은 최소 1000원 이상입니다.";
+	public static final String ERROR_WINNING_NUMBER_TYPE = "당첨 번호는 (, )를 구분자로 숫자만 입력 가능합니다. \n[예시] 1, 2, 3, 4, 5, 6";
```
상수와 관련해서 드린 코멘트와 다르게 이 부분도 함께 고민해야할 부분이 있을 것 같아요 :)
여기서 대부분의 메시지는 UI, 즉 View의 요구사항입니다.

`핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.`
위와 같은 규칙을 지켜보는건 어떨까요 ? View의 요구사항이 변경되었을 때 View와 관련된 모듈만 변경된다면 좋을 것 같아요.
응집도가 낮은 코드는 버그를 유발하기 쉽습니다 :) 응집도에 관련된 좋은 링크 남겨드릴게요 :)

https://madplay.github.io/post/coupling-and-cohesion-in-software-engineering

leeyohan93 commented on src/main/java/lotto/InputView.java:
```diff
@@ -0,0 +1,17 @@
+package lotto;
```
하나의 패키지에 lotto, ui 모델이 모두 구현되어 있는데요! 패키지를 나누어보는건 어떨까요 ?
패키지를 나누게 되면 모듈간의 의존성을 확인하기 쉽게 만들어 주고 가독성 좋은 구조로 만들어 줄 수 있어요 :)

leeyohan93 commented on src/main/java/lotto/LottoApplicationController.java:
```diff
@@ -0,0 +1,71 @@
+	public boolean enterPurchaseAmount() {
+		try {
+			purchaseAmount = new PurchaseAmount(InputView.enterPurchaseAmount());
+		} catch (IllegalArgumentException exception) {
+			OutputView.printMessage(exception.getMessage());
+			return true;
+		} catch (IllegalStateException exception) {
+			OutputView.printMessage(exception.getMessage());
+			return true;
+		}
```
요구사항 외의 기능 구현 👍  코드가 중복된 곳이 있네요 😄

다만, 만약에 View가 통제 가능한 콘솔이 아니라 웹 뷰라고 가정해본다면 어떨까요?
아래 인용문들과 함께 고민해보면 좋을 것 같아요!

이펙티브자바 아이템 75. 예외의 상세 메시지에 실패 관련 정보를 담으라
이펙티브자바 아이템 77. 예외를 무시하지말라

grow-with replied:
해당 코드는 IllegalArgumentException 과 IllegalStateException 일 떄 메세지를 사용자에게 출력하는 부분인데 중복되었다고 말씀하시는건 메세지의 중복을 의미하시는 걸까요?? 
그리고 View가 웹뷰일때 어떤 차이점이 존재하는지 제가 아직 잘 이해가 안가서 ㅠ 설명을 좀 부탁드려도 될까요?

leeyohan93 replied:
중복되었다고 말씀드린건 제가 잘못 보았네요 죄송합니다 😢 

Controller라는 네이밍을 사용하셨기에, 
만약 웹 뷰였다면 예외 상황 때 예외를 캐치해서 다음과 같이 반환하는 것이 적합할까 고민이 들었던 코멘트였습니다.
지금은 콘솔에 최적화 시켜주셨기 때문에 굳이 지금 단계에서는 바꾸지 않아도 괜찮을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/InputView.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.Scanner;
+
+public class InputView {
+	private static Scanner scanner = new Scanner(System.in);
```
final을 깜빡 하신 것 같아요  :)
```suggestion
	private static final Scanner scanner = new Scanner(System.in);
```

leeyohan93 commented on src/main/java/lotto/PurchaseAmount.java:
```diff
@@ -0,0 +1,29 @@
+package lotto;
+
+public class PurchaseAmount {
```
PurchaseAmount에 대한 테스트도 함께 작성해보는 건 어떨까요 ?

leeyohan93 commented on src/main/java/lotto/LottoApplicationController.java:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	public void printPurchaseQuantity() {
+		int purchaseQuantity = purchaseAmount.value() / CommonConstant.LOTTO_SALES_PRICE;
```
purchaseAmount의 상태를 확인하고 직접 계산하는 것보다 객체에게 메시지를 던져보는 것은 어떨까요 ?
```suggestion
		int purchaseQuantity = purchaseAmount.divide(CommonConstant.LOTTO_SALES_PRICE);
```

leeyohan93 commented on src/main/java/lotto/LottoApplicationController.java:
```diff
@@ -0,0 +1,71 @@
+		if (purchaseQuantity > CommonConstant.PURCHASE_FINISH) {
+			return true;
+		}
+		return false;
```
코드를 인라인으로 사용해보는 건 어떨까요 ? :)
```suggestion
		return purchaseQuantity > CommonConstant.PURCHASE_FINISH;
```

leeyohan93 commented on src/test/java/lotto/LottoNumberFactoryTest.java:
```diff
@@ -0,0 +1,33 @@
+		//given
+		List<Integer> numbers = LottoNumberFactory.create();
+		//when
+		List<Integer> distinctNumbers = numbers.stream().distinct().collect(Collectors.toList());
```
하나의 라인에 하나의 .을 사용하면 가독성이 좋아집니다 :)

```suggestion
		List<Integer> distinctNumbers = numbers.stream()
                .distinct()
                .collect(Collectors.toList());
```

leeyohan93 commented on src/main/java/lotto/LottoNumberFactory.java:
```diff
@@ -0,0 +1,22 @@
+		}
+	}
+
+	public static List<Integer> create() {
```
로또 번호를 생성하는 책임을 가지는 클래스를 도출하셨네요! 👍 

추가적으로 로또 번호(LottoNumber) 클래스를 도출해보는 것은 어떨까요 ?

> 규칙 3: 모든 원시값과 문자열을 포장한다.

leeyohan93 commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,31 @@
+
+	public String StringValues() {
+		return lottoNumbers.toString();
+	}
```
개인적으로는 toString 메서드를 View를 위해 사용하는건 적절하지 않다고 생각해요.

view의 요구사항이 바뀐다면 우리는 도메인 모델의 toString을 변경해주어야 하기 때문입니다.

만약 다른 개발자가 로깅을 위해서 lottoNumbers의 toString을 오버라이드한다면 어떻게 될까요? 🤔

leeyohan93 commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,31 @@
+		for (Integer lottoNumber : lottoNumbers) {
+			this.lottoNumbers.add(lottoNumber);
+		}
+	}
```
일급 컬렉션을 불변 객체와 같이 만들어보는 건 어떨까요 ?
Collection에서 제공하는 unmodifiableList를 활용하면 리스트에 대한 추가, 삭제를 막아주는 기능으로 일급 컬렉션도 좀 더 불변 객체에 가까워질 수 있습니다 :)

추가적으로 생성자를 사용할 때 주 생성자라는 개념이 있는데 참고해보시면 좋을 것 같아요 :)
https://jackjeong.tistory.com/148

```suggestion
    private final List<Integer> lottoNumbers;

    public LottoNumbers(int... lottoNumbers) {
        this(Arrays.stream(lottoNumbers)
                .boxed()
                .collect(Collectors.toList()));
    }

    public LottoNumbers(List<Integer> lottoNumbers) {
        this.lottoNumbers = lottoNumbers;
    }
```

leeyohan93 commented on src/main/java/lotto/WinnerCount.java:
```diff
@@ -0,0 +1,17 @@
+package lotto;
+
+public class WinnerCount {
+	private int winnerCount;
```
모든 원시 값을 포장하신 것이 좋았습니다 ! 👍 
다만, 클래스를 작게 분리하는 것은 정말 중요하지만, 적절한 책임에의해 만들어져야하며 항상 오버엔지니어링은 경계해야합니다. :) 해당 클래스는 어떤 책임을 담당할까요?

leeyohan93 commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,25 @@
+
+	public Lotto() {
+		this.lottoNumbers = new LottoNumbers(LottoNumberFactory.create());
+	}
```
지금 상황에서는 Lotto() 생성자를 호출하는 로직은 테스트가 어렵다는 단점이 있는데요.
LottoNumberFactory로 로또의 생성의 책임을 분리하였는데 내부에서 다시 합쳐진 것 같아요 
Lotto 내부에서 LottoNumberFactory를 호출하는 것보다 외부에서 호출하도록 분리해보면 어떨까요? :)

```suggestion
	public Lotto(LottoNumbers lottoNumbers) {
		this.lottoNumbers = new LottoNumbers(lottoNumbers);
	}
```

leeyohan93 commented on src/main/java/lotto/WinningRecord.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class WinningRecord {
+	private final LinkedHashMap<Integer, WinningInformation> winningRecord = new LinkedHashMap<>();
+
+	public WinningRecord() {
+		int[] ranks = {CommonConstant.FIFTH_PLACE, CommonConstant.FOURTH_PLACE,
+						CommonConstant.THIRD_PLACE, CommonConstant.FIRST_PLACE};
+		int[] winningAmount = {CommonConstant.FIFTH_AMOUNT, CommonConstant.FOURTH_AMOUNT,
+						CommonConstant.THIRD_AMOUNT, CommonConstant.FIRST_AMOUNT};
+		for (int i = CommonConstant.FIRST_INDEX; i < ranks.length; i++) {
+			winningRecord.put(ranks[i],
+				new WinningInformation(ranks[i], winningAmount[i], CommonConstant.DEFAULT_VALUE));
+		}
+	}
```
순서에 의존하는 코드는 실수하기 쉽고 어렵습니다.
로또 티켓의 당첨 결과를 의미하는 Enum을 만들어보면 어떨까요? 

다음의 두 책임을 하나로 묶는 클래스가 만들어질 수 있을 것 같아요 :)
- 조건에 맞는 등수를 구할 수 있다.
- 등수 별 당첨 금액을 구할 수 있다.

leeyohan93 reviewed (changes requested):
책임에 대해 많이 고민하시고 클래스를 나누려는 노력이 코드에서 보였습니다 👍
몇가지 코멘트들을 남겼으니 확인 부탁드릴게요 :)

주말에도 열심히 미션 수행하시는 모습에 저도 열심히 남겨보았습니다! 😄 
궁금한 점은 DM이나 코멘트 남겨주세요 :)

grow-with committed:
refactor: 낮은 응집도를 위해 기존의 공통 상수, 메세지 클래스 제거

grow-with committed:
refactor: 클래스간 구분을 위해 패키지 생성

grow-with committed:
refactor: 누락된 final 선언

grow-with committed:
test: PurchaseAmount 테스트 코드 작성

grow-with committed:
refactor: purchaseAmount 코드 수정 및 테스트 작성

- purchase() 메서드 추가
- purchase() 메서드 관련 테스트 코드 작성
- 패키지 생성으로 테스트코드 패키지 변경

grow-with committed:
refactor: continuePurchase() 인라인으로 코드 수정

grow-with committed:
refactor: 하나의 라인에 하나의 .만 사용하도록 수정

grow-with committed:
refactor: view에 전달하는 값으로 toString() 사용안하도록 수정

grow-with committed:
refactor: 일급 컬렉션 불변객체로 수정, 주 생성자 패턴으로 수정

grow-with committed:
refactor: Lotto클래스, LottoNumberFactory 의존성 해제

- 생성자에서 LottoNumberFactory.create() 부분 외부로 이동
- 주 생성자를 가지도록 생성자 코드 수정

grow-with committed:
refactor: isWinner() 코드 인라인으로 변경

grow-with committed:
refactor: 기존의 당첨통계에 사용하던 클래스 전체 수정

- Enum 클래스 Rank 생성 (일치번호 개수, 당첨금액)
- 당첨자 수를 기록하는 일급 컬렉션 Winners 생성
- Winners 테스트 코드 작성

grow-with committed:
refactor: controller에서 view 의존성 제거

- view를 main메서드 안에서 호출하도록 의존성 분리

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,25 @@
+public class Lotto {
+	private final LottoNumbers lottoNumbers;
+
+	public Lotto(LottoNumbers lottoNumbers) {
+		this.lottoNumbers = lottoNumbers;
+	}
+
+	public Lotto(int... lottoNumbers) {
+		this(new LottoNumbers(lottoNumbers));
+	}
```
주 생성자의 위치를 부 생성자 아래에 위치해보는건 어떨까요?

https://jackjeong.tistory.com/148

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,25 @@
+		return count;
+	}
+
+	public String getLottoNumbersStringValues() {
+		return lottoNumbers.StringValues();
+	}
```
https://github.com/next-step/java-lotto-pro/pull/170#discussion_r744193082

위 코멘트와 이어지는 부분인데요 :)

`핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.`
위와 같은 규칙을 지켜보는건 어떨까요 ? View의 요구사항이 변경되었을 때 View와 관련된 모듈만 변경된다면 좋을 것 같아요.
응집도가 낮은 코드는 버그를 유발하기 쉽습니다 :) 응집도에 관련된 좋은 링크 남겨드릴게요 :)

지금 도메인 레이어 전반적으로 View에 대한 요구사항들이 구현되어있는데 다음 단계 시작하시기 전에 모두 분리해보면 좋을 것 같아요 :)

https://madplay.github.io/post/coupling-and-cohesion-in-software-engineering

grow-with replied:
안녕하세요 항상 좋은 코드리뷰 감사합니다.
제가 코드수정 방향에 대해 조금 확신이 안들어서 질문을 몇가지 드릴려고 합니다 ㅠ
도메인레이어에 view의 요구사항이 반영되어있다 라고 하셨는데 이게 도메인레이어에서 view가 원하는 서식 혹은 포맷의 값을 준다 라고 해석하면 될까요?
어찌됫든 도메인에서 값을 뽑아서 뷰에 전달을 해야 출력을 할 수 있을거라고 생각을 해서 이해가 조금 잘 안됩니다 ㅠ
아니면 혹시 뷰에서 도메인이 접근해서 값을 출력하는 식으로 분리하는걸 얘기하신걸까요??

leeyohan93 replied:
> 도메인레이어에 view의 요구사항이 반영되어있다 라고 하셨는데 이게 도메인레이어에서 view가 원하는 서식 혹은 포맷의 값을 준다 라고 해석하면 될까요?

네 맞습니다. 지금 현재 `return lottoNumbers.StringValues();` 메서드 내부에서 view에서 원하는 서식으로 맞춰주고 있지요.
여기서 만약 View의 서식이 변한다면, 저희는 도메인 모델의 로직을 수정해야합니다. 
이것은 view와 관련된 로직이 view에 응집 되어있지 않고 도메인 모델에 구현되어 있기 때문인데요.

이 문제를 해결하기 위해서는 도메인 모델에 view의 서식을 맞추는 로직을 모두 제거하고 getter를 통해 내부 값만 반환을 한 후에 view모델 안에서 서식을 맞추는 것입니다. 이렇게 되면 view의 요구사항이 변경되더라도 도메인 로직에 대한 변경이 하나도 이루어지지 않도록 만들 수 있습니다. :)

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,39 @@
+	private static final int MATCHED_NUMBER_PLUS = 1;
+	private static final int NON_MATCHED_NUMBER = 0;
+	private static final String DELIMITER = ", ";
+	private final List<Integer> lottoNumbers;
```
https://github.com/next-step/java-lotto-pro/pull/170#discussion_r744192806

grow-with replied:
lottonumber 의미를 가진 클래스로 저는 lottonumbers를 만든건데 단일 번호를 가지는 lottonumber클래스를 의미하시는걸까요?

leeyohan93 replied:
네 맞습니다 Integer 타입을 LottoNumber 클래스로 래핑하는 것에 대한 코멘트입니다 :)

leeyohan93 commented on src/main/java/lotto/domain/LottoNumberFactory.java:
```diff
@@ -0,0 +1,26 @@
+public class LottoNumberFactory {
+	private static final int FIRST_INDEX = 0;
+	private static final int SIX_INDEX = 6;
+	private static final int MIN_LOTTO_NUMBER = 1;
+	private static final int MAX_LOTTO_NUMBER = 45;
```
LottoNumber 클래스가 생긴다면 더 적합한 위치가 될 수 있겠네요 🤔

grow-with replied:
팩토리 클래스에서 로또넘버들을 생성해주기 위한 상수값인데 LottoNumber클래스에 적합한 위치일 것이다라는 부분이 잘 이해가 가지 않습니다 ㅠ
저는 LottoNumbers라는 클래스를 로또번호 6자리를 가지는 의미로 만들어준 것인데 LottoNumber 클래스는 1자리의 번호를 의미하시는 걸까요??

leeyohan93 replied:
로또 번호(LottoNumber)의 정의는 무엇일까요 ? 제가 생각하는 로또번호의 정의는 1~45 사이의 숫자입니다.
상수를 배치할 때는 가장 관련있는 클래스에 배치하는 것이 좋다고 생각하여 위에서 코멘트드린 로또 번호(LottoNumber) 클래스를 만들어주신다면 그 내부에 배치하면 좋을 것 같다고 생가갛여 코멘트 드린 부분입니다 :)

leeyohan93 commented on src/main/java/lotto/domain/LottoNumberFactory.java:
```diff
@@ -0,0 +1,26 @@
+
+public class LottoNumberFactory {
+	private static final int FIRST_INDEX = 0;
+	private static final int SIX_INDEX = 6;
```
조금 더 의미 있는 네이밍을 지을 수도 있을 것 같아요 :)
만약 다른 자리 수의 숫자의 로또를 생성한다고 가정해보면, 해당 상수 명도 바뀌어야합니다

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,39 @@
+		this.lottoNumbers = Collections.unmodifiableList(new ArrayList<>(lottoNumbers));
+	}
+
+	public int isMatch(int winningNumber) {
```
isMatch라는 네이밍의 함수라면 리턴 값으로 boolean을 리턴할 것이라고 기대하게 되는데 어떻게 생각하시나요 ? 🤔

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,39 @@
+		return NON_MATCHED_NUMBER;
+	}
+
+	public String StringValues() {
```
메서드 명은 항상 소문자로 시작해야합니다 :)

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LottoNumbers {
```
`6개의 로또 번호를 가져야한다` 라는 요구사항에 대한 검증이 빠진 것 같습니다!

grow-with replied:
6개의 로또 번호 검증을 LottoNumberFactoryTest에서 작성했는데 LottoNumbers 테스트코드를 추가적으로 작성해야 할까요??

leeyohan93 replied:
네 추가적으로 작성해주시는 것이 좋습니다 :)
오히려 검증의 책임을 LottoNumbers로 가져오는 것이 더욱 적합할 수도 있습니다.
LottoNumberFactoryTest에서 작성해주셨다 하더라도 
LottoNumbers는 항상 6개의 로또 번호를 가진다는 100% 신뢰가 가능할까요? 🤔 
먼역 LottoNumbers를 LottoNumberFactory가 아닌 다른 곳에서 생성하게 되면 어떻게 될까요?

일급 컬렉션의 장점 중 비즈니스에 종속적인 자료구조를 만들 수 있다는 것입니다.
지금의 구조에서는 LottoNumbers를 100% 신뢰하기는 힘들어보이는 것 같아요 :)

https://jojoldu.tistory.com/412

leeyohan93 commented on src/main/java/lotto/domain/LottoStatistics.java:
```diff
@@ -0,0 +1,28 @@
+package lotto.domain;
+
+public class LottoStatistics {
+	private static final int LOTTO_PRICE = 1000;
```
로또 한줄의 가격을 의미하는 LOTTO_PRICE는 LottoNumbers 클래스 내부에 있는 것이 더욱 적합하지 않을까요?
상수에 대한 공유가 필요할때는 접근제어자를 public으로 수정해서 사용해주세요 :)

grow-with replied:
LottoNumbers클래스는 로또번호를 저장하는 역할을 담당하구 LottoStatistics클래스는 수익률을 구하기 위해서 해당 상수값을 가지고 있습니다.
제 생각에는 책임에 따라서 잘 분리했다고 생각했는데 리뷰해주신 부분의 의미를 제가 잘 파악을 못하고 있어서 ㅠ 조금 더 상세하게 설명해주실 수 있을까요?

leeyohan93 replied:
로또의 가격이 바뀐다고 가정해보시면 이해가 더욱 되실 것 같습니다 !
만약 로또의 가격이 2000원이 되었다면 어떻게 될까요? 우리는 같은 의미의 상수를 여러 곳에서 정의해놓았기 때문에
변경 범위가 넓어지게 되고 이는 버그를 유발하기 쉬운 구조가 됩니다.
같은 의미의 상수라면 public으로 변경하고 함께 공유하는 것이 좋지 않을까요? 🤔

leeyohan93 commented on src/main/java/lotto/domain/MatchedNumber.java:
```diff
@@ -0,0 +1,18 @@
+package lotto.domain;
+
+public class MatchedNumber {
```
사용하지 않는 클래스는 제거해주세요 :)

leeyohan93 commented on src/main/java/lotto/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,36 @@
+public class PurchaseAmount {
+	private static final String ERROR_PURCHASE_AMOUNT_TYPE = "[ERROR] 구매금액은 숫자만 입력가능합니다.";
+	private static final String ERROR_PURCHASE_AMOUNT = "[ERROR] 구매금액은 최소 1000원 이상입니다.";
+	private static final int LOTTO_PRICE = 1000;
```
중복된 상수를 제거해주시면 좋을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/Quantity.java:
```diff
@@ -0,0 +1,13 @@
+package lotto.domain;
+
+public class Quantity {
```
https://github.com/next-step/java-lotto-pro/pull/170#discussion_r744193792

leeyohan93 commented on src/main/java/lotto/domain/Winners.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class Winners {
```
로또의 결과를 구하는 로직이 LottoStatistics과 분리되어 있는데 조금으 오버엔지니어링 된 느낌인 것 같습니다 🤔 
Winners 는 데이터를 저장하기 위한 클래스처럼 보이기도 하는데요, 
이 두 클래스를 적절히 합쳐서 더욱 적절한 책임을 가진 클래스를 만들어보는건 어떨까요?

leeyohan93 commented on src/main/java/lotto/domain/Winners.java:
```diff
@@ -0,0 +1,20 @@
+public class Winners {
+	private static final int DEFAULT_VALUE = 0;
+	private static final int ADD_VALUE = 1;
+	private final Map<Integer, Integer> winners = new HashMap<>();
```
```suggestion
	private final Map<Rank, Integer> winners = new EnumMap<>(Rank.class);
```

EnumMap을 홀용해볼 수도 있을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,38 @@
+	private static final String ERROR_WINNING_NUMBER_TYPE = "당첨 번호는 (, )를 구분자로 숫자만 입력 가능합니다. \n[예시] 1, 2, 3, 4, 5, 6";
+	private static final String DELIMITER = ", ";
+	private static final int MIN_REQUIRED_NUMBER_LOTTO = 6;
+	private final List<Integer> winningNumbers = new ArrayList<>();
```
화면, 뷰의 요구사항을 제외한 WinningNumbers와 LottoNumbers가 가지는 차이점이 있을까요? 
당첨 번호 또한 로또 번호이니 LottoNumbers를 사용해보는건 어떨까요 ? 🤔

leeyohan93 commented on src/main/java/lotto/domain/LottoNumberFactory.java:
```diff
@@ -0,0 +1,26 @@
+		}
+	}
+
+	public static List<Integer> create() {
```
```suggestion
	public static Lotto create() {
```

로또를 생성하는 기능을 가졌으니 Lotto 클래스의 일급컬렉션을 만들어 Lottos를 반환해보는건 어떨까요?
클래스 이름도 수정이 필요할 수도 있을 것 같네요 :)

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -0,0 +1,64 @@
+	private PurchaseAmount purchaseAmount;
+	private WinningNumbers winningNumbers;
+	private LottoStatistics lottoStatistics = new LottoStatistics();
+	private List<Lotto> lottos = new ArrayList<>();
```
```suggestion
	private Lottos lottos;
```
Lotto에 대한 일급 컬렉션을 만들어보는 건 어떨까요 ?

추가로 코멘트 드려보자면 전반적으로 클래스 내부에 다른 클래스에 대한 참조가 있을 때 , 
미리 초기화를 해주고 난 후에 상태를 변경하는 방식으로 구현되어 있는데요. 
불변 객체로 사용하기 위한 리팩터링을 해보시는건 어떨까요 !? :)

https://velog.io/@conatuseus/Java-Immutable-Object%EB%B6%88%EB%B3%80%EA%B0%9D%EC%B2%B4

leeyohan93 reviewed (approved):
피드백 반영해주신 부분 잘 보았습니다 ! 👍 
몇가지 피드백 들을 남겼으니 확인 부탁드릴게요 :)

이번 단계는 원활한 진행을 위해 머지 하겠습니다. 😄 
다음 단계 시작 전에 코멘트 들에 대해 고민해보시고 궁금한 점이 있으시다면 언제든 코멘트, DM 남겨주세요 👍  남은 단계도 화이팅입니다 🔥
